src/Schema/Directives/GuardDirective.php

```diff
@@ -0,0 +1,126 @@
+<?php
+
+namespace Nuwave\Lighthouse\Schema\Directives;
+
+use Closure;
+use GraphQL\Language\AST\TypeDefinitionNode;
+use GraphQL\Language\AST\TypeExtensionNode;
+use Illuminate\Contracts\Auth\Factory as AuthFactory;
+use Nuwave\Lighthouse\Exceptions\AuthenticationException;
+use Nuwave\Lighthouse\Schema\AST\ASTHelper;
+use Nuwave\Lighthouse\Schema\AST\DocumentAST;
+use Nuwave\Lighthouse\Schema\Values\FieldValue;
+use Nuwave\Lighthouse\Support\Contracts\DefinedDirective;
+use Nuwave\Lighthouse\Support\Contracts\FieldMiddleware;
+use Nuwave\Lighthouse\Support\Contracts\TypeExtensionManipulator;
+use Nuwave\Lighthouse\Support\Contracts\TypeManipulator;
+
+/**
+ * @see \Illuminate\Auth\Middleware\Authenticate
+ */
+class GuardDirective extends BaseDirective implements FieldMiddleware, TypeManipulator, TypeExtensionManipulator, DefinedDirective
+{
+    /**
+     * @var \Illuminate\Contracts\Auth\Factory
+     */
+    protected $auth;
+
+    /**
+     * @param  \Illuminate\Contracts\Auth\Factory  $auth
+     */
+    public function __construct(AuthFactory $auth)
+    {
+        $this->auth = $auth;
+    }
+
+    public static function definition(): string
+    {
+        return /* @lang GraphQL */ <<<'SDL'
+"""
+Run authentication through one more guards.
+This is run per field and may allow unauthenticated
+users to still receive partial results.
+"""
+directive @guard(
+  """
+  Specify which guards to use, e.g. "api".
+  When not defined, the default driver is used.
+  """
+  with: [String!]
+) on FIELD_DEFINITION | OBJECT
+SDL;
+    }
+
+    /**
+     * Resolve the field directive.
+     *
+     * @param  \Nuwave\Lighthouse\Schema\Values\FieldValue  $fieldValue
+     * @param  \Closure  $next
+     * @return \Nuwave\Lighthouse\Schema\Values\FieldValue
+     */
+    public function handleField(FieldValue $fieldValue, Closure $next): FieldValue
+    {
+        $resolver = $fieldValue->getResolver();
+
+        return $next(
+            $fieldValue->setResolver(
+                function () use ($resolver) {
+                    $this->authenticate(
+                        (array) $this->directiveArgValue('with')
+                    );
+
+                    return $resolver(...func_get_args());
+                }
+            )
+        );
```

Comment:
TBO, I don't understand this peace of code... I am neither fit with Laravel authorization, nor with some lighthouse core functionality.

Reply:
The basic structure here is what all field middleware do: Wrap around the previous resolver and do some steps inbetween. The call to `authenticate` will throw in case the user is not authenticated.

I guess the usage of the magic `func_get_args()` method is a bit confusing. It is in there because we do not care about the resolver arguments at all in this middleware, so we just pass them through without even naming them specifically.